### ISSUE TITLE
Improve build times for remote layers

### ISSUE THREAD

**Issue by dsyer** [kubernetes-sigs/kustomize#2460]
There is plenty of evidence that build times are too long when remote layers are used (#1132, #1431 and #1147). The issues keep getting closed, so apparently no-one cares. But why? If it takes more than a minute to build a simple deployment manifest that's too long IMO. A cache would help a  lot - it seems crazy to keep cloning the same github repo over and over, which seems to be what happens.

**Comment by monopole** [comment]
fwiw, see #2579

The immediate solution to slow build times in the presence of `resource:` entries that are specified as git urls is to 1) clone the git repo locally, 2) if there's no kustomization file in the repo, drop one in to the local copy (so that kustomize recognizes it as a 'legal' source of config), and 3) and change the `resource:` entry from a git URL to a relative or absolute file path on the user's  local file system.

Obviously this uses git itself as the `cache` manager with a human in the loop.  It's safe, because the user's copy of the repo necessarily has a _reflog_ and is in a well-defined state with respect to tags, and the user can always rebase to capture upstream changes.  That's always been the [intended workflow](https://github.com/kubernetes-sigs/kustomize/blob/master/docs/workflows.md#2-clone-it-as-your-base).

Right now the git cloner clone to a temp directory foe the user as a convenience, then wipes it post build - because everyone would have a different opinion of where this local, temporary copy (i.e.  _cache_) should be kept, or whether or not it should be kept at all.  The ability to do this (use a git url and clone it to a temp dir) was conceived as a convenience for _developing_ a config pipeline as a POC, not as a day-to-day thing to do over and over.  There was always an assumption that one would maintain a local clone

And now that [`kpt`](https://googlecontainertools.github.io/kpt) exists, that's an even better way to maintain a base.  It's a git repo + k8s specific metadata.

leaving this open for discussion. another option is deprecate the notion of allowing git urls in resources, to avoid the appearance that it's a recommended practice.

**Comment by pst** [comment]
If builds (in general) should depend on external sources is an endless discussion with people on both sides. I really only see two options:

 1. Either Kustomize supports remote bases, then the feature also needs to be maintained and work properly. 
 2. Or it does not, in which case at least there is no confusion.

I've previously contributed git remote handling that did reduce the amount of data transferred to speed up builds and keep build times largely independent of the repository growth. This unfortunately has been reverted to a more naive approach when there was an issue with it. Whatever the reason was, I still think this is a valid approach to investigate. From what I researched at the time, this is the same logic most CI/CD systems implement.

Alternatively, of course there is caching. Freedesktop has [`$XDG_CACHE_HOME`](https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html) which could be used to have a sane default for caches while giving people control if necessary.

Personally I do not use remote bases, I think vendoring bases is the better approach (e.g. my [Terraform framework with Kustomize integration](https://www.kubestack.com/catalog)), because it prevents the risk of my build failing due to external sources being unavailable. I've also chose not to incluce remote bases support in the [Terraform Kustomize provider](https://github.com/kbst/terraform-provider-kustomize/issues/10) I built because the feature has a troubled history upstream.

Frankly, if somebody steps up to fix remote bases, let them fix it. If not, maybe it's better to just remove it altogether.

**Comment by vaidik** [comment]
Is there a recommended way to do vendoring of bases other than manually doing it using git submodules? Always hard to reason about for most people. And as far as I know, there is no native way in kustomize to do this. If there is though, I would also like that to be an option.

**Comment by dsyer** [comment]
Bump

**Comment by bajtos** [comment]
Build times are still too long when remote layers are used :(

**Comment by maknihamdi** [comment]
could this PR  https://github.com/kubernetes-sigs/kustomize/pull/3655 fixes the issue?

**Comment by maknihamdi** [comment]
We use kustomize with my team to manage specs, we use a gitops approach with a single repo entry-point, and a multilayer customizations, using remote references to git repositories. With this approach, kustomize build can take 15 minutes and more, depending on environments (development, production,..) and components.

in the same 'kustomize build', the same repository with the same ref (tag or commit id) can be used many times. With the actual implementation of cloner this repo is cloned as many times as its visited.
Because of kustromize is the tool we use many times a day, we loose a lot of time waiting kustomize command cloning our repositories many times a day, this is way we decides to fork the project and use our custom kustomize with some optimizations:

1. The first 'evident' think to do is to stop cloning the same think n times in the same build cycle. the git ref is a tag, branch or commit id.
2. between 2 build command, we should reuse the cloned copy if the reference is immutable: commit id, or a tag (I know tag can be removed and recreated, user should clean manually his cache in this case), otherwise we should clean the clone 

So we have 2 requirements:
1. a cache in the same kustomize command execution,
2. a cache between 2 different kustomize command executions

How we do that:
1. We stop reference every remote resource in a random directory name. a RepoSpec of a repository git.com/repo?ref=myRef should have the same Dir path.
2. A new function CachedGitCloner can be used by the loader to give him a cached copy of a repository if the repoSpec.Dir exists
3. We still clone the repository in /tmp/ksutomize-* directory and then move it to the cache directory.
4. the CachedGitCloner checks if the repoSpec.ref is a mutable reference, if it is a branch we move the repo directory to a subdirectory "branches", user can clean the cache manually later.

As we discuss with @monopole here https://github.com/kubernetes-sigs/kustomize/pull/3655 , we can improve this workflow by introducing a cache expiration, and/or adding a new command or flag to activate or not the git cache, another one to clean the cache

**Comment by james-callahan** [comment]
@monopole could you reopen this issue? it seems to be the one with the most discussion on the issue.

We'd love to see it addressed; quite possibly via #3655

**Comment by max-sixty** [comment]
I recognize the auto-close bot can be useful to prune issues that don't have a reasonable chance of making forward progress.

But is it working as intended here? Is the message of the maintainers that this is a non-starter? All good if so — but if not, maybe we could ask the auto-close bot to relax a little?

**Comment by ringerc** [comment]
@max-sixty Right. This doesn't seem to be fixed or resolved in any way, nor has a decision not to do it been made.

Kustomize doesn't have a way to cache remote layers. Nor does it have a way to accept a map of git `--reference` repos for faster cloning, or a way to rewrite or remap remote layers to local URIs.

**Comment by ringerc** [comment]
In addition to build times, this is a maintenance mess.

URIs to other repos get embedded, creating long chains of references. `bases` is a string-list, so you can't separate the remote `ref` from the URI string, making it hard to handle version overrides, remapping and lookups.

There's no easy way recursively clone the remotes and rewrite the kustomizations to point to local paths.

**Comment by ringerc** [comment]
See also https://github.com/kubernetes-sigs/kustomize/issues/1431

**Comment by koba1t** [comment]
/reopen
details: https://github.com/kubernetes-sigs/kustomize/pull/3655#issuecomment-3010240466

**Comment by ringerc** [comment]
Another argument for having this is that `kustomize` _does_ cache Helm charts in its Helm expander.

It does so by adding the cached charts to a directory in the main repo tree, where they may be optionally `git add`ed and versioned. This gets a bit ugly, but can be mostly handled with appropriate use of git's `':!*/charts/*` filters when using `git diff`, `git grep` etc. This has the advantage that the artefacts a given kustomization depends on are now versioned along with the kustomizations themselves.

**Comment by Jaakkonen** [comment]
(Commenting as this is the issue you find from search engines when looking for kustomize caching)

Just noting that this has been implemented for Helm charts which automatically pull the helm charts to `charts/` folder and raw URL resources can be cached using `kustomize localize`. Source: https://github.com/kubernetes-sigs/kustomize/blob/master/kustomize/commands/localize/localize.go
